// @christianjuth/ts-cli-generator/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "build", {
    enumerable: true,
    get: function() {
        return _build.build;
    }
});
Object.defineProperty(exports, "run", {
    enumerable: true,
    get: function() {
        return _run.run;
    }
});
Object.defineProperty(exports, "call", {
    enumerable: true,
    get: function() {
        return _run.call;
    }
});
Object.defineProperty(exports, "CLI", {
    enumerable: true,
    get: function() {
        return _types.CLI;
    }
});
Object.defineProperty(exports, "Events", {
    enumerable: true,
    get: function() {
        return _types.Events;
    }
});
Object.defineProperty(exports, "credentials", {
    enumerable: true,
    get: function() {
        return _auth.credentials;
    }
});
Object.defineProperty(exports, "formatTable", {
    enumerable: true,
    get: function() {
        return _utils.formatTable;
    }
});
Object.defineProperty(exports, "printTable", {
    enumerable: true,
    get: function() {
        return _utils.printTable;
    }
});
Object.defineProperty(exports, "setRoot", {
    enumerable: true,
    get: function() {
        return _config.setRoot;
    }
});
var _build = require("./build");
var _run = require("./run");
var _types = require("./types");
var _auth = require("./auth");
var _utils = require("./utils");
var _config = require("./config");
